### ISSUE TITLE
MIPS MSB, the compilation journey

### ISSUE THREAD

**Issue by FlUxIuS** [frida/frida#1615]
Hi, 

It seems the support for MIPS is not finished looking at ticket [#152](https://github.com/frida/frida/issues/152). But maybe these steps will improve the installation that is a bit of a journey and I was still stuck yesterday evening at the end to compile everything.

So first, I compile the SDK as follows:

```
make -f Makefile.sdk.mk FRIDA_HOST=linux-mips FRIDA_LIBC=gnu
[...]
╭────
│ 📦 libffi
├───────────────────────────────────────────────╮
│ URL: https://github.com/frida/libffi.git
│ CID: 10d9cbec8b4537b8d1452c73f7651e95d22d0a17
├───────────────────────────────────────────────╯
│ libffi :: Cloning into deps/libffi
│ libffi :: Building
failed - see build/fs-tmp-linux-mips/libffi/build.log for more information
make: *** [Makefile.sdk.mk:112: build/fs-linux-mips/manifest/libffi.pkg] Error 1
```

Looking at the `build.log` it seems the generated `meson.build` uses an non-intialized variable for MIPS case:

```
cat build/fs-tmp-linux-mips/libffi/build.log
[...]
Compiler for C supports function attribute visibility: YES 
Program test-cc-supports-hidden-visibility.py found: YES (/usr/bin/python3 /fridapath/deps/libffi/test-cc-supports-hidden-visibility.py)
Message: .hidden pseudo-op is available
Program msvcc.sh found: YES (/fridapath/deps/libffi/msvcc.sh)

deps/libffi/src/meson.build:70:2: ERROR: Unknown variable "c_sources".

A full log can be found at /fridapath/build/fs-tmp-linux-mips/libffi/meson-logs/meson-log.txt
```

Quick direct fix:
```
$ diff deps/libffi/src/meson.build.old deps/libffi/src/meson.build                                                                                                                                           
70,71c70,71
<   c_sources += ['ffi.c']
<   asm_sources += ['o32.S', 'n32.S']
---
>   c_sources = ['ffi.c']
>   asm_sources = ['o32.S', 'n32.S'
```

We continue the compilation:
```
make -f Makefile.sdk.mk FRIDA_HOST=linux-mips FRIDA_LIBC=gnu
[...]
╭────
│ 📦 quickjs
├───────────────────────────────────────────────╮
│ URL: https://github.com/frida/quickjs.git
│ CID: 3026500048050c6e29ea376dcd8207af27a25d13
├───────────────────────────────────────────────╯
│ quickjs :: Cloning into deps/quickjs
│ quickjs :: Building
failed - see build/fs-tmp-linux-mips/quickjs/build.log for more information
make: *** [Makefile.sdk.mk:112: build/fs-linux-mips/manifest/quickjs.pkg] Error 
```

So here some definintions are missing:
```
cat build/fs-tmp-linux-mips/quickjs/build.log
[...]
mips-unknown-linux-gnu-gcc -Ilibquickjs.a.p -I. -I../../../deps/quickjs -I/fridapath/build/fs-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Os -g -include config.h -ffunction-sections -fdata-sections -march=mips1 -fPIC -MD -MQ libquickjs.a.p/quickjs.c.o -MF libquickjs.a.p/quickjs.c.o.d -o libquickjs.a.p/quickjs.c.o -c ../../../deps/quickjs/quickjs.c
../../../deps/quickjs/quickjs.c: In function 'js_ecvt':
../../../deps/quickjs/quickjs.c:11494:66: error: 'FE_DOWNWARD' undeclared (first use in this function)
                 js_ecvt1(d, n_digits + 1, &decpt1, &sign1, buf1, FE_DOWNWARD,
                                                                  ^~~~~~~~~~~
../../../deps/quickjs/quickjs.c:11494:66: note: each undeclared identifier is reported only once for each function it appears in
../../../deps/quickjs/quickjs.c:11496:66: error: 'FE_UPWARD' undeclared (first use in this function)
                 js_ecvt1(d, n_digits + 1, &decpt2, &sign2, buf2, FE_UPWARD,
                                                                  ^~~~~~~~~
../../../deps/quickjs/quickjs.c: In function 'js_fcvt':
../../../deps/quickjs/quickjs.c:11545:64: error: 'FE_DOWNWARD' undeclared (first use in this function)
             n1 = js_fcvt1(buf1, sizeof(buf1), d, n_digits + 1, FE_DOWNWARD);
                                                                ^~~~~~~~~~~
../../../deps/quickjs/quickjs.c:11546:64: error: 'FE_UPWARD' undeclared (first use in this function)
             n2 = js_fcvt1(buf2, sizeof(buf2), d, n_digits + 1, FE_UPWARD);
                                                                ^~~~~~~~~
ninja: build stopped: subcommand failed.
```

It's bad I know! But I quickly fixed it including these definitions to continue the compilation:
```
$ diff deps/quickjs/quickjs.c.old deps/quickjs/quickjs.c                                                                                                                                                     2 

49a50,52
> #define FE_DOWNWARD 0x3
> #define FE_UPWARD   0x2
> 
```

Then, a new problem with openssl:
```
make -f Makefile.sdk.mk FRIDA_HOST=linux-mips FRIDA_LIBC=gnu
[...]
╭────
│ 📦 openssl 1.1.1i
├───────
│ URL: https://www.openssl.org/source/openssl-1.1.1i.tar.gz
│ SHA: e8be6a35fe41d10603c3cc635e93289ed00bf34b79671a3a4de64fcee00d5242
├────────────
│ openssl :: Downloading
│ openssl :: Verifying
│ openssl :: Extracting to deps/openssl
│ openssl :: Applying openssl-clang.patch
│ openssl :: Applying openssl-android.patch
│ openssl :: Applying openssl-armcap.patch
│ openssl :: Setting up build directory
│ openssl :: Building
make: *** [Makefile.sdk.mk:337: build/fs-linux-mips/manifest/openssl.pkg] Error 2
```

Seems like there is a confusion with `-mips2` and `-mips1` flags:
```
cat build/fs-tmp-linux-mips/openssl/build.log
[...]
make[2]: Entering directory '/fridapath/build/fs-tmp-linux-mips/openssl'
make[2]: Leaving directory '/fridapath/build/fs-tmp-linux-mips/openssl'
make[2]: Entering directory '/fridapath/build/fs-tmp-linux-mips/openssl'
mips-unknown-linux-gnu-gcc -static-libgcc  -I. -Iinclude -fPIC -pthread -mabi=32 -mips2 -Wa,--noexecstack -fPIC -ffunction-sections -fdata-sections -march=mips1 -Os -pipe -g1 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BN_ASM_MONT -DSHA1_ASM -DSHA256_ASM -DAES_ASM -DOPENSSLDIR="\"/etc/ssl\"" -DENGINESDIR="\"/fridapath/build/fs-linux-mips/lib/engines-1.1\"" -DNDEBUG -I/fridapath/build/fs-linux-mips/include -MMD -MF apps/app_rand.d.tmp -MT apps/app_rand.o -c -o apps/app_rand.o apps/app_rand.c
Assembler messages:
Error: -mips2 conflicts with the other architecture options, which imply -mips1
apps/app_rand.c:1:0: error: '-mips2' conflicts with the other architecture options, which specify a mips1 processor
 /*
 
make[2]: *** [Makefile:670: apps/app_rand.o] Error 1
make[2]: Leaving directory '/fridapath/build/fs-tmp-linux-mips/openssl'
make[1]: *** [Makefile:177: build_libs] Error 2
make[1]: Leaving directory '/fridapath/build/fs-tmp-linux-mips/openssl'
```

Here is my dirty quick fix:
```
Configure:1246:        $value = '-mips1' if ($target =~ /mips32/);
Configurations/10-main.conf:103:          /^MIPS16/     && do { $wcecdefs.=" -DMIPS -D_MIPS_ -DR4000 -D$wcetgt";
Configurations/10-main.conf:104:                                $wcecdefs.=" -DMIPSII -QMmips16";
Configurations/10-main.conf:105:                                $wcelflag.=" /machine:MIPS16";  last; };
Configurations/10-main.conf:107:                                $wcecdefs.=" -QMmips1";
Configurations/10-main.conf:1672:        cflags           => "-mrtp -mips1 -B\$(WIND_BASE)/host/\$(WIND_HOST_TYPE)/lib/gcc-lib/ -msoft-float -mno-branch-likely -fno-builtin -fno-defer-pop",
crypto/bn/asm/mips.pl:92:	$code="#if !(defined (__mips_isa_rev) && (__mips_isa_rev >= 6))\n.set     mips1\n#endif\n";
```

At the end we have a SDK well compiled, but if I try to compile gum, we have new issues:
```
$ make gum-linux-mips FRIDA_LIBC=gnu                                                                                                                                                                       130 ↵
make[1]: Entering directory '/fridapath'
. build/frida_thin-meson-env-linux-mips.rc; builddir=build/tmp_thin-linux-mips/frida-gum; if [ ! -f $builddir/build.ninja ]; then python3 /fridapath/releng/meson/meson.py --cross-file build/frida_thin-linux-mips.txt --prefix /fridapath/build/frida_thin-linux-mips --libdir /fridapath/build/frida_thin-linux-mips/lib --default-library static -Doptimization=s -Db_ndebug=true --strip -Dv8=enabled frida-gum $builddir || exit 1; fi; /fridapath/releng/ninja-linux-x86_64 -C $builddir install || exit 1
ninja: Entering directory `build/tmp_thin-linux-mips/frida-gum'
[1/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8thread.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8thread.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8thread.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8thread.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8thread.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8thread.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.h:13,
                 from ../../../frida-gum/bindings/gumjs/gumv8core.h:12,
                 from ../../../frida-gum/bindings/gumjs/gumv8thread.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8thread.cpp:7:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[2/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8value.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8value.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8value.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8value.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8value.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8value.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.h:13,
                 from ../../../frida-gum/bindings/gumjs/gumv8core.h:12,
                 from ../../../frida-gum/bindings/gumjs/gumv8value.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8value.cpp:7:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[3/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8bundle.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8bundle.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8bundle.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8bundle.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8bundle.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8bundle.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8bundle.h:11,
                 from ../../../frida-gum/bindings/gumjs/gumv8bundle.cpp:7:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[4/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8memory.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8memory.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8memory.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8memory.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8memory.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8memory.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.h:13,
                 from ../../../frida-gum/bindings/gumjs/gumv8core.h:12,
                 from ../../../frida-gum/bindings/gumjs/gumv8memory.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8memory.cpp:7:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[5/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8stream.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8stream.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8stream.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8stream.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8stream.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8stream.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.h:13,
                 from ../../../frida-gum/bindings/gumjs/gumv8core.h:12,
                 from ../../../frida-gum/bindings/gumjs/gumv8object.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8stream.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8stream.cpp:7:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[6/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8object.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8object.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8object.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8object.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8object.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8object.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.h:13,
                 from ../../../frida-gum/bindings/gumjs/gumv8core.h:12,
                 from ../../../frida-gum/bindings/gumjs/gumv8object.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8object.cpp:7:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[7/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scope.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scope.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scope.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scope.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scope.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8scope.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.h:13,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.cpp:8:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[8/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8script.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8script.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8script.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8script.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8script.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8script.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8scope.h:13,
                 from ../../../frida-gum/bindings/gumjs/gumv8core.h:12,
                 from ../../../frida-gum/bindings/gumjs/gumv8object.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8apiresolver.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8script-priv.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8script.cpp:11:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[9/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scriptbackend.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scriptbackend.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scriptbackend.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scriptbackend.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8scriptbackend.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8scriptbackend.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8bundle.h:11,
                 from ../../../frida-gum/bindings/gumjs/gumv8platform.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8scriptbackend.cpp:12:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
[10/29] Compiling C++ object bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8platform.cpp.o
FAILED: bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8platform.cpp.o 
mips-unknown-linux-glibc-g++ -Ibindings/gumjs/libfrida-gumjs-1.0.a.p -Ibindings/gumjs -I../../../frida-gum/bindings/gumjs -Ibindings -I../../../frida-gum/bindings -I. -I../../../frida-gum -Igum -I../../../frida-gum/gum -I../../../frida-gum/gum/arch-x86 -I../../../frida-gum/gum/arch-arm -I../../../frida-gum/gum/arch-arm64 -I../../../frida-gum/gum/arch-mips -Ilibs -I../../../frida-gum/libs -Ilibs/gum/heap -I../../../frida-gum/libs/gum/heap -Ilibs/gum/prof -I../../../frida-gum/libs/gum/prof -I../../../frida-gum/gum/backend-elf -I/fridapath/build/frida_thin-sdk-linux-mips/include/glib-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/lib/glib-2.0/include -I/fridapath/build/frida_thin-sdk-linux-mips/include -I/fridapath/build/frida_thin-sdk-linux-mips/include/capstone -I/fridapath/build/frida_thin-sdk-linux-mips/include/quickjs -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8 -I/fridapath/build/frida_thin-sdk-linux-mips/include/gio-unix-2.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/json-glib-1.0 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libsoup-2.4 -I/fridapath/build/frida_thin-sdk-linux-mips/include/libxml2 -I/fridapath/build/frida_thin-linux-mips/include -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -std=c++14 -fno-exceptions -fno-rtti -Os -g -include config.h '-DG_LOG_DOMAIN="Frida"' -D_GNU_SOURCE=1 -ffunction-sections -fdata-sections -march=mips1 -include /fridapath/build/frida-version.h -fPIC -pthread -MD -MQ bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8platform.cpp.o -MF bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8platform.cpp.o.d -o bindings/gumjs/libfrida-gumjs-1.0.a.p/gumv8platform.cpp.o -c ../../../frida-gum/bindings/gumjs/gumv8platform.cpp
In file included from /fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8.h:30:0,
                 from ../../../frida-gum/bindings/gumjs/gumv8bundle.h:11,
                 from ../../../frida-gum/bindings/gumjs/gumv8platform.h:10,
                 from ../../../frida-gum/bindings/gumjs/gumv8platform.cpp:7:
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:101:1: error: static assertion failed: Pointer compression can be enabled only for 64-bit architectures
 static_assert(
 ^~~~~~~~~~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:71: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                                       ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:412:68: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprHeapReservationSize = size_t{1} << 32;
                                                          ~~~~~~~~~~^~~~~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:72: warning: left shift count >= width of type [-Wshift-count-overflow]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                                        ^~
/fridapath/build/frida_thin-sdk-linux-mips/include/v8-8.0/v8/v8-internal.h:413:69: error: right operand of shift expression '(1u << 32)' is >= than the precision of the left operand [-fpermissive]
   static constexpr size_t kPtrComprIsolateRootAlignment = size_t{1} << 32;
                                                           ~~~~~~~~~~^~~~~
ninja: build stopped: subcommand failed.
make[1]: *** [Makefile.linux.mk:119: build/frida_thin-linux-mips/lib/pkgconfig/frida-gum-1.0.pc] Error 1
make[1]: Leaving directory '/fridapath'
make: *** [Makefile:4: gum-linux-mips] Error 2
```
And then it's too late! I gave up last night, but I'll try to continue or I hope other people will have the motivation to do so.

**Comment by oleavr** [comment]
Thanks for digging into this!

- libffi: Just updated our patches to properly fix the MIPS and S390x build system bits. (Made a mistake when rebasing those recently.)
- QuickJS: Compiles fine on Ubuntu 20.04's cross-compiler, so I suspect we'll need a compatibility kludge for older libc versions. (Your workaround should work fine for most Frida agents -- it's only really an issue if using some computational JS library that relies on rounding behavior.)
- OpenSSL: The issue is that the MIPS32 assembly code requires `-mips2`, according to the comment in `Configure`. I think the correct fix here is to disable the assembly bits when targeting MIPS. Just fixed this in `Makefile.sdk.mk`.
- V8: It seems like V8 was configured to use pointer compression, but this doesn't make sense on 32-bit architectures. Probably easy to fix, but an easy workaround is to disable V8 in `config.mk` by changing `FRIDA_V8` there to `disabled`.

By the way, I've just tweaked our build environment so `FRIDA_LIBC=gnu` is the default, and also made sure we default to a toolprefix matching Ubuntu's cross compiler.

Note to self: It would be great to add CI for this soon to ensure this doesn't regress in the future.

**Comment by oleavr** [comment]
Forgot to mention that I added `-mfp32` to the arch flags, as GCC complained that this is required for `-march=mips1`. Wonder if this would actually fix the QuickJS build issue for you?

**Comment by oleavr** [comment]
The issue with V8 was that I had forgotten to wire up the MIPS build system bits, so it was actually built for your build machine (x86_64?) instead of MIPS. Fixed this in e1168ae9ff75d8438a30e969b252cd8e200971ec.

**Comment by oleavr** [comment]
Turns out V8 has some (trivial) MIPS regressions which I've just fixed and will submit upstream. I've bumped `v8_version` in `releng/deps.mk` but if you don't want to nuke your local checkout the easiest is probably: `cd deps/v8-checkout/v8 && git checkout master && git pull`. Just make sure you use the `clean-v8` target of `Makefile.sdk.mk` before building it again.

**Comment by oleavr** [comment]
Note that for building V8 for big-endian MIPS you will need binfmt_misc hooked up to QEMU so V8's build system can run the `mksnapshot` binary built for MIPS. (Or you can tweak Makefile.sdk.mk to build V8 with snapshots disabled, at the expense of startup time.)

**Comment by FlUxIuS** [comment]
Nice fixes thanks! :)

Is there a clean way to set the path of shared libraries the `./mksnapshot` when run with QEMU thanks to `binfmt_misc`?

```
FAILED: gen/embedded.S gen/snapshot.cc 
python ../../../deps/v8-checkout/v8/tools/run.py ./mksnapshot --turbo_instruction_scheduling --target_os=linux --target_arch=mips --embedded_src gen/embedded.S --embedded_variant Default --random-seed 314159265 --startup_src gen/snapshot.cc --no-native-code-counters
/lib/ld.so.1: No such file or directory

```

**Comment by FlUxIuS** [comment]
Or maybe doing an independent snapshot in a chroot dir?

**Comment by oleavr** [comment]
@FlUxIuS It's done like this:

```sh
$ export QEMU_LD_PREFIX=/usr/mips-linux-gnu
```

I just tried this and there's another issue: that mksnapshot tries to reserve 2 GiB of address space, which won't fly on 32-bit MIPS. I've narrowed it down to this code in `src/snapshot/mksnapshot.cc`:

```c++
      // Set code range such that relative jumps for builtins to
      // builtin calls in the snapshot are possible.
      i::Isolate* i_isolate = reinterpret_cast<i::Isolate*>(isolate);
      size_t code_range_size_mb =
          i::kMaximalCodeRangeSize == 0
              ? i::kMaxPCRelativeCodeRangeInMB
              : std::min(i::kMaximalCodeRangeSize / i::MB,
                         i::kMaxPCRelativeCodeRangeInMB);
      v8::ResourceConstraints constraints;
      constraints.set_code_range_size_in_bytes(code_range_size_mb * i::MB);
      i_isolate->heap()->ConfigureHeap(constraints);
```

Should be an easy fix, let's give this a few minutes of thought.

**Comment by FlUxIuS** [comment]
> ```shell
> export QEMU_LD_PREFIX=/usr/mips-linux-gnu
> ```

Oh I was doing complicated things! Thanks for that :)

**Comment by oleavr** [comment]
@FlUxIuS I just learned about it myself :)

This change seems related to the issue
https://github.com/v8/v8/commit/27a71e94f9f4fda5c2626627f40c07c5d93275bf

**Comment by oleavr** [comment]
@FlUxIuS I think the solution is to lower `kMaximalCodeRangeSize` when the code being compiled is MIPS-native. (E.g. when building for MIPSEL it will be x86-native as the snapshot tool is run natively, because endian matches there there's no need for running a MIPS binary.)

**Comment by FlUxIuS** [comment]
So probably we should add a patch to `releng` when MIPS is chosen

**Comment by oleavr** [comment]
@FlUxIuS We have a fork of the V8 repo in our GH organization to keep track of patches. I just pushed a [fix](https://github.com/frida/v8/commit/35fc430315daa11147829be927663ff049f0b149) there and bumped `v8_version` in `releng/deps.mk`. (So do like above to update your V8 without starting fresh.) Builds fine for me now, but have no idea if it works :)

**Comment by FlUxIuS** [comment]
Nice thanks! I'll try this weekend and come back to complain =P

**Comment by FlUxIuS** [comment]
Hey,

Continuing the journey and with a fresh Linux setup everything compiles perfectly with the SDK! Good job! 

About the GUM part, I tried to compile and do few quick fixes, but there are some dependencies to resolve still.

So first, I got an error when compiling gum with the following command: 
```
$ make build/frida-linux-mips/lib/pkgconfig/frida-gum-1.0.pc
[...]
../../../frida-gum/bindings/gumjs/gumv8process.cpp: In function ‘void _gum_v8_process_init(GumV8Process*, GumV8Module*, GumV8Core*, v8::Local<v8::ObjectTemplate>)’:
../../../frida-gum/bindings/gumjs/gumv8process.cpp:119:44: error: ‘GUM_SCRIPT_ARCH’ was not declared in this scope; did you mean ‘GUM_SCRIPT’?
  119 |       String::NewFromUtf8Literal (isolate, GUM_SCRIPT_ARCH), ReadOnly);
      |                                            ^~~~~~~~~~~~~~~
      |                                            GUM_SCRIPT
```

So I've quickly fixed it adding the mission definitions as follows:

```
$ diff ./frida-gum/bindings/gumjs/gumv8process.cpp.old ./frida-gum/bindings/gumjs/gumv8process.cpp
30a31,32
> #elif defined (HAVE_MIPS)
> # define GUM_SCRIPT_ARCH "mips"
```

And then I got other errors due to missing libraries: 
[buildgum.log](https://github.com/frida/frida/files/5866642/buildgum.log)

So I've fixed the first one with the missing atomic library:

```
$ diff meson.build.old meson.build                                  
241c241
<     extra_libs_private += ['-ldwarf', '-ldl', '-lz']
---
>     extra_libs_private += ['-ldwarf', '-ldl', '-lz', '-latomic']
```

But there are still others to resolve like tinycc, etc. I tried to quickly add `-ltiny`, but still having the same issue for the moment.

Cheers,

**Comment by oleavr** [comment]
@FlUxIuS Just pushed some fixes, so at least Gum is building now. You will have to rebuild the SDK unfortunately. (The packaging has been updated to omit TinyCC on architectures it doesn't support, and fixing Libs.private in V8's .pc.) Next up seems to be missing labrats binaries in frida-core.

Note to self: MIPS CI coverage is badly needed, considering that the port isn't that old and yet we'd already accumulated quite a few regressions :sweat_smile:
